qiskit_ibm_provider/transpiler/passes/scheduling/block_base_padder.py

```diff
@@ -184,11 +219,216 @@ def _pad(
+
+    def _visit_if_else_op(self, node: DAGNode) -> None:
+        """check if is fast-path eligible otherwise fall back
+        to standard ControlFlowOp handling."""
+
+        if self._will_use_fast_path(node):
+            self._fast_path_nodes.add(node)
+        self._visit_control_flow_op(node)
+
+    def _will_use_fast_path(self, node: DAGNode) -> bool:
+        """Check if this conditional operation will be scheduled on the fastpath.
+        This will happen if
+        1. This operation is a direct descendent of a current measurement block to be flushed
+        2. The operation only operates on the qubit that is measured.
+        """
+        # Verify IfElseOp has a direct measurement predecessor
+        condition_bits = node.op.condition_bits
+        for bit in condition_bits:
+            last_node = self._last_node_to_touch.get(bit, None)
+
+            last_node_in_block = True
+            # TODO: find way to check if node in DAG without using private attribute.
+            if last_node is not None:
+                last_node_in_block = True
+                try:
+                    self._block_dag.node(last_node._node_id)
+                except IndexError:
+                    last_node_in_block = False
+            else:
+                last_node_in_block = False
+
+            if not (
+                last_node_in_block
+                and isinstance(last_node.op, Measure)
+                and set(node.qargs) == set(last_node.qargs)
+            ):
+                return False
+
+        # Fast path contents are limited to gates and delays
+        for block in node.op.blocks:
+            block_dag = circuit_to_dag(block)
+            for node_ in block_dag.topological_op_nodes():
+                if not isinstance(node_.op, (Gate, Delay)):
+                    return False
+        return True
+
+    def _visit_control_flow_op(self, node: DAGNode) -> None:
+        """Visit a control-flow node to pad."""
+
+        # Control-flow terminator ends scheduling of block currently
+        block_idx, t0 = self._node_start_time[node]  # pylint: disable=invalid-name
+        self._terminate_block(t0, block_idx)
+        self._add_block_terminating_barrier(block_idx, t0, node)
+
+        # Only pad non-fast path nodes
+        fast_path_node = node in self._fast_path_nodes
+
+        # TODO: This is a hack required to tie nodes of control-flow
+        # blocks across the scheduler and block_base_padder. This is
+        # because the current control flow nodes store the block as a
+        # circuit which is not hashable. For processing we are currently
+        # required to convert each circuit block to a dag which is inefficient
+        # and causes node relationships stored in analysis to be lost between
+        # passes as we are constantly recreating the block dags.
+        # We resolve this here by extracting the cached dag blocks that were
+        # stored by the scheduling pass.
```

Comment:
Circuits never will be hashable, so if your logic wants that, you'll likely need to find another way of communicating between the passes.  The point about needing to swap back and forth between DAG and `QuantumCircuit` is a known issue, though - it's a big part of the "new IR" discussion.  For the initial releases of control-flow support in the transpiler, we just said "screw it" and ate the extra cost of more than one pass converting back-and-forth, since we didn't have time to design the new IR immediately.

Reply:
I did find a different way of doing this (how the two passes are working together right now), but its a note for the future. Ideally, I would pass an IR from pass to pass to mutate without having to reconstruct a completely new IR.